### PR TITLE
Make server timeout configurable

### PR DESCRIPTION
**Pull request by mikaelm12**
Issue: https://github.com/aspnet/AspNetCore/issues/6797

Making the server side disconnect timeout configurable.

### CODE REVIEW

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs [hunk 1]
```diff
@@ -31,12 +32,14 @@ public partial class HttpConnectionManager
         private readonly ILogger<HttpConnectionManager> _logger;
         private readonly ILogger<HttpConnectionContext> _connectionLogger;
 
-        public HttpConnectionManager(ILoggerFactory loggerFactory, IApplicationLifetime appLifetime)
+        public TimeSpan DisconnectTimeout { get; set; }
+
+        public HttpConnectionManager(ILoggerFactory loggerFactory, IApplicationLifetime appLifetime, IOptions<ConnectionOptions> connectionOptions)
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs]: This is a breaking change but I'm not sure how much we care. Add another constructor until we decide it's OK to break here?

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs [hunk 2]
```diff
@@ -34,17 +34,23 @@ public partial class HttpConnectionManager
...
             _nextHeartbeat = new TimerAwaitable(_heartbeatTickRate, _heartbeatTickRate);
-            DisconnectTimeout = connectionOptions.Value.DisconnectTimeout ?? ConnectionOptionsSetup.DefaultDisconectTimeout;
+            DisconnectTimeout = ConnectionOptionsSetup.DefaultDisconectTimeout;
             // Register these last as the callbacks could run immediately
             appLifetime.ApplicationStarted.Register(() => Start());
             appLifetime.ApplicationStopping.Register(() => CloseConnections());
         }
 
+        public HttpConnectionManager(ILoggerFactory loggerFactory, IApplicationLifetime appLifetime, IOptions<ConnectionOptions> connectionOptions)
+            :this(loggerFactory, appLifetime)
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs]: nit: formatting... 😄

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs [hunk 3]
```diff
@@ -31,17 +32,25 @@ public partial class HttpConnectionManager
...
             _connectionLogger = loggerFactory.CreateLogger<HttpConnectionContext>();
             _nextHeartbeat = new TimerAwaitable(_heartbeatTickRate, _heartbeatTickRate);
-
+            DisconnectTimeout = ConnectionOptionsSetup.DefaultDisconectTimeout;
             // Register these last as the callbacks could run immediately
             appLifetime.ApplicationStarted.Register(() => Start());
             appLifetime.ApplicationStopping.Register(() => CloseConnections());
         }
 
+        public HttpConnectionManager(ILoggerFactory loggerFactory, IApplicationLifetime appLifetime, IOptions<ConnectionOptions> connectionOptions)
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionManager.cs]: Usually, the one with the most argument does the actual work and the ones with less arguments just delegate to that one.

File: src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs [hunk 4]
```diff
@@ -439,7 +440,7 @@ public async Task TransportEndingGracefullyWaitsOnApplicationLongPolling()
                     await task.OrTimeout();
 
                     // We've been gone longer than the expiration time
-                    connection.LastSeenUtc = DateTime.UtcNow.Subtract(TimeSpan.FromSeconds(10));
+                    connection.LastSeenUtc = DateTime.UtcNow.Subtract(TimeSpan.FromSeconds(20));
```
Comment (BrennanConroy) [inline src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs]: We should probably explicitly set the DisconnectTimeout in this test

File: src/SignalR/common/Http.Connections/src/ConnectionsDependencyInjectionExtensions.cs [hunk 5]
```diff
@@ -20,9 +23,22 @@ public static IServiceCollection AddConnections(this IServiceCollection services
...
             return services;
         }
+
+        /// <summary>
+        /// Adds required services for ASP.NET Core Connection Handlers to the specified <see cref="IServiceCollection" />.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection" /> to add services to.</param>
+        /// <param name="options">A callback to configure  <see cref="ConnectionOptions" /></param>
+        /// <returns>The same instance of the <see cref="IServiceCollection"/> for chaining.</returns>
+        public static IServiceCollection AddConnections(this IServiceCollection services, Action<ConnectionOptions> options)
```
Comment (BrennanConroy) [inline src/SignalR/common/Http.Connections/src/ConnectionsDependencyInjectionExtensions.cs]: What happens if you call the AddConnections overload with options after AddSignalR? And if you call the overload before AddSignalR?

Comment (mikaelm12) [reply src/SignalR/common/Http.Connections/src/ConnectionsDependencyInjectionExtensions.cs]: I tried the variations on the SignalRSamples project and it still works, propagating the passed in values and uses the default properly.

**Comment by mikaelm12** [comment]
🆙 📅

**Review by davidfowl — APPROVED** [review]

**Pull request merged**